### PR TITLE
Support abstract classes declaring static interned instances of their own type

### PR DESCRIPTION
**Pull request by Stephan202**
Summary says it.

### CODE REVIEW

File: value-fixture/src/org/immutables/fixture/InternedClassWithSelfConstant.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+   limitations under the License.
+ */
+package org.immutables.fixture;
+
+import org.immutables.value.Value;
+
+@Value.Immutable(intern = true)
+@Value.Style(allParameters = true)
+abstract class InternedClassWithSelfConstant {
+  public static final ImmutableInternedClassWithSelfConstant FOO = ImmutableInternedClassWithSelfConstant.of("foo");
```
Comment (Stephan202) [inline value-fixture/src/org/immutables/fixture/InternedClassWithSelfConstant.java:23]: This is what the PR is about: the changes make this construction possible.

File: value-fixture/test/org/immutables/fixture/InternedClassWithSelfConstantTest.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+ */
+package org.immutables.fixture;
+
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.assertSame;
+
+public class InternedClassWithSelfConstantTest {
+  @Test
+  public void testInstantiation() {
+    assertSame(ImmutableInternedClassWithSelfConstant.of("foo"), ImmutableInternedClassWithSelfConstant.FOO);
```
Comment (Stephan202) [inline value-fixture/test/org/immutables/fixture/InternedClassWithSelfConstantTest.java:24]: Without the fix below this test fails with:
```
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.017 s <<< FAILURE! - in org.immutables.fixture.InternedClassWithSelfConstantTest
[ERROR] testInstantiation  Time elapsed: 0.015 s  <<< ERROR!
java.lang.ExceptionInInitializerError
	at org.immutables.fixture.InternedClassWithSelfConstantTest.testInstantiation(InternedClassWithSelfConstantTest.java:24)
Caused by: java.lang.NullPointerException
	at org.immutables.fixture.InternedClassWithSelfConstantTest.testInstantiation(InternedClassWithSelfConstantTest.java:24)
```

File: value-processor/src/org/immutables/value/processor/Immutables.generator [hunk 3]
```diff
@@ -697,21 +697,23 @@ public [type.typeAbstract.relative] [v.names.with]([v.atNullability][v.type] val
 [else if type.useSingletonOnly][-- deliberately left empty --]
 [else if type.useInterned]
 
+  private static final class InternerHolder {
```
Comment (Stephan202) [inline value-processor/src/org/immutables/value/processor/Immutables.generator:700]: We're effectively applying the [Initialization-on-demand holder idiom](https://en.wikipedia.org/wiki/Initialization-on-demand_holder_idiom).

**Review by Stephan202 — COMMENTED** [review]
Some context :)

File: value-processor/src/org/immutables/value/processor/Immutables.generator [hunk 4]
```diff
@@ -697,21 +697,23 @@ public [type.typeAbstract.relative] [v.names.with]([v.atNullability][v.type] val
...
   [if type.generateJdkOnly]
     [if type.useWeakInterned]
-  private static final java.util.Map<[type.typeImmutable.simple], java.lang.ref.WeakReference<[type.typeImmutable.simple]>> INTERNER =
-      new java.util.WeakHashMap<>();
+    static final java.util.Map<[type.typeImmutable.simple], java.lang.ref.WeakReference<[type.typeImmutable.simple]>> INTERNER =
+        new java.util.WeakHashMap<>();
     [else]
-  private static final java.util.concurrent.ConcurrentHashMap<InternProxy, [type.typeImmutable.relative]> INTERNER =
+    static final java.util.concurrent.ConcurrentHashMap<InternProxy, [type.typeImmutable.relative]> INTERNER =
```
Comment (elucash) [inline value-processor/src/org/immutables/value/processor/Immutables.generator:706]: I'm merging this, just in case what removing `private` does here

**Comment by elucash** [comment]
Thanks!

**Pull request merged**